### PR TITLE
Cpp/feature/basic connection

### PR DESCRIPTION
**Pull request by Shylock-Hg**
Follow https://github.com/vesoft-inc/nebula-common/pull/295

### CODE REVIEW

File: cpp/src/client/ClientImpl.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
+/* Copyright (c) 2020 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include <memory>
+
+#include "./ClientImpl.h"
```
Comment (laura-ding) [inline cpp/src/client/ClientImpl.cpp]: use whole path

File: cpp/src/client/ClientImpl.cpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+namespace nebula {
+
+bool ClientImpl::open(const std::string &address, int32_t port, uint32_t timeout) {
+    auto socket = apache::thrift::async::TAsyncSocket::newSocket(
```
Comment (laura-ding) [inline cpp/src/client/ClientImpl.cpp]: try catch

File: cpp/src/client/ClientImpl.cpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+AuthResponse ClientImpl::authenticate(const std::string &user, const std::string &password) {
+    if (client_ == nullptr) {
+        return AuthResponse{ErrorCode::E_DISCONNECTED};
+    }
+
+    graph::cpp2::AuthResponse resp;
+    try {
+        client_->sync_authenticate(resp, user, password);
+    } catch (const std::exception &ex) {
+        return AuthResponse{ErrorCode::E_RPC_FAILURE, -1};
```
Comment (laura-ding) [inline cpp/src/client/ClientImpl.cpp]: need to output the error msg

File: cpp/src/client/ClientImpl.cpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+void ClientImpl::signout(int64_t sessionId) {
+    if (client_ != nullptr) {
+        client_->sync_signout(sessionId);
+    }
+}
+
+/*static*/ ErrorCode ClientImpl::from(graph::cpp2::ErrorCode code) {
+    switch (code) {
```
Comment (laura-ding) [inline cpp/src/client/ClientImpl.cpp]: @dutor it needs to packing data? You said didn't need it

File: cpp/src/client/ClientImpl.cpp [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+#include <memory>
+
+#include "./ClientImpl.h"
+
+namespace nebula {
+
+bool ClientImpl::open(const std::string &address, int32_t port, uint32_t timeout) {
+    auto socket = apache::thrift::async::TAsyncSocket::newSocket(
+        folly::EventBaseManager::get()->getEventBase(), address, port, timeout);
```
Comment (laura-ding) [inline cpp/src/client/ClientImpl.cpp]: To confirm the timeout if need check

File: cpp/src/client/ClientImpl.cpp [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+    client_ = std::make_unique<graph::cpp2::GraphServiceAsyncClient>(
+        apache::thrift::HeaderClientChannel::newChannel(socket));
+    return true;
+}
+
+void ClientImpl::close() {
+    static_cast<apache::thrift::ClientChannel*>(client_->getChannel())->closeNow();
```
Comment (laura-ding) [inline cpp/src/client/ClientImpl.cpp]: check the client_ is no null

File: cpp/src/client/ClientImpl.cpp (hunk 5 above)
Comment (Shylock-Hg) [reply cpp/src/client/ClientImpl.cpp]: zero is ok

File: cpp/src/client/ClientImpl.cpp (hunk 6 above)
Comment (Shylock-Hg) [reply cpp/src/client/ClientImpl.cpp]: Good catch. Fixed.

File: cpp/src/client/ClientImpl.cpp (hunk 1 above)
Comment (Shylock-Hg) [reply cpp/src/client/ClientImpl.cpp]: This header only used in library.

File: cpp/src/client/ClientImpl.cpp (hunk 2 above)
Comment (Shylock-Hg) [reply cpp/src/client/ClientImpl.cpp]: Fixed.

File: cpp/src/client/ClientImpl.cpp (hunk 3 above)
Comment (Shylock-Hg) [reply cpp/src/client/ClientImpl.cpp]: Where to output?

Comment (laura-ding) [reply cpp/src/client/ClientImpl.cpp]: the AuthResponse has the error_msg

File: cpp/src/client/ClientImpl.cpp (hunk 4 above)
Comment (laura-ding) [reply cpp/src/client/ClientImpl.cpp]: I think the ErrorCode and ExecutionResponse can warpper in common repo.

File: cpp/src/client/ClientImpl.cpp (hunk 3 above)
Comment (Shylock-Hg) [reply cpp/src/client/ClientImpl.cpp]: OK

File: cpp/examples/RawConnection.cpp [hunk 7]
```diff
@@ -0,0 +1,32 @@
+/* Copyright (c) 2020 vesoft inc. All rights reserved.
```
Comment (laura-ding) [inline cpp/examples/RawConnection.cpp]: The users couldn't use the connection directly.  the example is for session and pool

Comment (Shylock-Hg) [reply cpp/examples/RawConnection.cpp]: I will provide the example for session later.

Comment (laura-ding) [reply cpp/examples/RawConnection.cpp]: so you can delete it

Comment (Shylock-Hg) [reply cpp/examples/RawConnection.cpp]: Ok, fixed.

File: cpp/src/client/ClientImpl.cpp [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        return ExecutionResponse{ErrorCode::E_RPC_FAILURE, 0};
+    }
+
+    return resp;
+}
+
+void ClientImpl::asyncExecute(int64_t sessionId,
+                              const std::string &stmt,
+                              Connection::ExecuteCallback cb) {
+    client_->future_execute(sessionId, stmt).thenValue([cb = std::move(cb)](auto &&resp) {
```
Comment (jievince) [inline cpp/src/client/ClientImpl.cpp]: Why some functions check if client_ is nullptr while some functions not?

**Review by jievince — DISMISSED** [review]

File: cpp/src/client/ClientImpl.cpp (hunk 8 above)
Comment (Shylock-Hg) [reply cpp/src/client/ClientImpl.cpp]: Because no better way to report error.

Comment (Shylock-Hg) [reply cpp/src/client/ClientImpl.cpp]: Please review the depended PR.

**Review by jievince — DISMISSED** [review]

**Pull request closed without merging**